test/unit/atomic_iteration_test.lua

```diff
@@ -0,0 +1,165 @@
+        for request_number, old_tuple, new_tuple, space_id in iterator() do
+            table.insert(transaction, old_tuple:totable())
+        end
+        table.insert(transactions, transaction)
+    end
+
+    local true_box_begin = box.begin
+
+    -- mock box.begin
+    box.begin = function ()
+        true_box_begin()
+        box.on_commit(f)
+    end
+
+    for _, space in pairs({g.tree, g.hash}) do
+        -- tuples expired in one atomic_iteration
+        space:insert({1, "3"})
+        space:insert({2, "2"})
+        space:insert({3, "1"})
+
+
+        local task = expirationd.start("clean_all", space.id, helpers.is_expired_debug,
+                {atomic_iteration = true})
+        -- wait for tuples expired
+        helpers.retrying({}, function()
+            if space.index[0].type == "HASH" then
+                t.assert_equals(helpers.iteration_result, {{3, "1"}, {2, "2"}, {1, "3"}})
+            else
+                t.assert_equals(helpers.iteration_result, {{1, "3"}, {2, "2"}, {3, "1"}})
+            end
+        end)
+        task:kill()
+        helpers.iteration_result = {}
+
+        -- check out three row transaction
+        if space.index[0].type == "HASH" then
+            t.assert_equals(transactions, {
+                { {3, "1"}, {2, "2"}, {1, "3"} }
+            })
+        else
+            t.assert_equals(transactions, {
+                { {1, "3"}, {2, "2"}, {3, "1"} }
+            })
+        end
+        transactions = {}
+
+        -- tuples expired in two atomic_iteration
+        space:insert({1, "3"})
+        space:insert({2, "2"})
+        space:insert({3, "1"})
+
+        task = expirationd.start("clean_all", space.id, helpers.is_expired_debug,
+                {atomic_iteration = true, tuples_per_iteration = 2})
+        -- wait for tuples expired
+        -- 2 seconds because suspend will be yield in task
+        helpers.retrying({}, function()
+            if space.index[0].type == "HASH" then
+                t.assert_equals(helpers.iteration_result, {{3, "1"}, {2, "2"}, {1, "3"}})
+            else
+                t.assert_equals(helpers.iteration_result, {{1, "3"}, {2, "2"}, {3, "1"}})
+            end
+        end)
+        task:kill()
+        helpers.iteration_result = {}
+
+        if space.index[0].type == "HASH" then
+            t.assert_equals(transactions, {
+                { {3, "1"}, {2, "2"} }, -- check two row transaction
+                { {1, "3"} }            -- check single row transactions
+            })
+        else
+            t.assert_equals(transactions, {
+                { {1, "3"}, {2, "2"} }, -- check two row transaction
+                { {3, "1"} }            -- check single row transactions
+            })
+        end
+
+        transactions = {}
+    end
+
+    -- unmock
+    box.begin = true_box_begin
+end
+
+function g.test_mvcc_vinyl_tx_conflict()
+    for i = 1,10 do
+        g.vinyl:insert({i, tostring(i), nil, nil, 0})
+    end
```

Comment:
Probably it would be great to add comment that we don't check Tarantool or vinyl as engine. We check that expiration task continue work even in case of conflict.